src/main/java/apoc/uuid/UuidHandler.java

```diff
@@ -0,0 +1,183 @@
+            " Set 'apoc.uuid.enabled=true' in your neo4j.conf file located in the $NEO4J_HOME/conf/ directory.";
+
+    UuidHandler(GraphDatabaseAPI api, Log log) {
+        properties = api.getDependencyResolver().resolveDependency(EmbeddedProxySPI.class).newGraphPropertiesProxy();
+        this.log = log;
+    }
+
+    private static void checkEnabled() {
+        if (properties == null) {
+            throw new RuntimeException(NOT_ENABLED_ERROR);
+        }
+    }
+
+    public static Map<String, Object> add(String label, Map<String, Object> params) {
+        checkEnabled();
+        checkConstraintUuid(label);
+        Map<String, Object> stringObjectMap = updateUuid(label, map("params", params));
+        if (toBoolean(params.getOrDefault("addToAlreadyPresent", true))) {
+            GraphDatabaseService db = properties.getGraphDatabase();
+            try (Transaction tx = db.beginTx()) {
+                db.findNodes(Label.label(label)).stream().filter(node -> !node.hasProperty(UUID_PROPERTY)).forEach(node -> {
+                    String uuid = UUID.randomUUID().toString();
+                    node.setProperty(UUID_PROPERTY, uuid);
+                });
+                tx.success();
+            }
+        }
+        return stringObjectMap;
+    }
+
+    private synchronized static Map<String, Object> updateUuid(String label, Map<String, Object> value) {
+        checkEnabled();
+        try (Transaction tx = properties.getGraphDatabase().beginTx()) {
+            uuid.clear();
+            String uuidProperty = (String) properties.getProperty(APOC_UUID, "{}");
+            uuid.putAll(Util.fromJson(uuidProperty, Map.class));
+            Map<String, Object> previous = null;
+            if (label != null) {
+                previous = (value == null) ? uuid.remove(label) : uuid.put(label, value);
+                if (value != null || previous != null) {
+                    properties.setProperty(APOC_UUID, Util.toJson(uuid));
+                }
+            }
+            tx.success();
+            return previous;
+        }
+    }
+
+    @Override
+    public Object beforeCommit(TransactionData txData) throws Exception {
+        if (uuid.containsKey("")) updateUuid(null,null);
+        GraphDatabaseService db = properties.getGraphDatabase();
+        Map<String, String> exceptions = new LinkedHashMap<>();
+        Iterable<Node> createdNodes = txData.createdNodes();
+        Iterable<PropertyEntry<Node>> assignedNodeProperties = txData.assignedNodeProperties();
+        Iterable<PropertyEntry<Node>> removedNodeProperties = txData.removedNodeProperties();
+
+        uuid.forEach((label, config) -> {
+            try (Transaction tx = db.beginTx()) {
+                if (createdNodes.iterator().hasNext()) {
+                    createdNodes.forEach(node -> {
+                        boolean containsLabel = containsLabel(label, node.getLabels());
+                        if (containsLabel && !node.hasProperty(UUID_PROPERTY)) {
```

Comment:
In case of the `uuid` property is set or removed we restore it to the previous `uuid` value:

https://github.com/neo4j-contrib/neo4j-apoc-procedures/blob/1fb474beb3d1a9640652cc22a916a7fcf9d8948c/src/main/java/apoc/uuid/UuidHandler.java#L117-L132

Reply:
@jexp A third case is that if the node is created with the uuid like this:
`CREATE (n:Test {name:'test', uuid:'dab404ee-391d-11e9-b210-d663bd873d93'})`
we use the that uuid, do you think we should replace it with one generated by us?